In [1]:
import pandas as pd
import numpy as np

data1 = pd.read_csv("dirty_iris.csv")
data1

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
2             6.2          NaN           5.4          2.3   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
..            ...          ...           ...          ...         ...
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa
148           6.4          3.1           NaN          1.8   virginica
149           5.8          2.6           4.0          NaN  versicolor

[150 rows x 5 columns]

### 1 identifying missing value 

In [8]:
data1.columns

Index(['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species'],
      dtype='object')

In [11]:
data2 = data1[['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species']]
data2.dtypes

Sepal.Length    float64
Sepal.Width     float64
Petal.Length    float64
Petal.Width     float64
Species          object
dtype: object

**data type of column seems to be valid in above**

In [2]:
data1.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0           6.4          3.2           4.5          1.5  versicolor
1           6.3          3.3           6.0          2.5   virginica
2           6.2          NaN           5.4          2.3   virginica
3           5.0          3.4           1.6          0.4      setosa
4           5.7          2.6           3.5          1.0  versicolor

**suspect found in head**

In [5]:
data1.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    140.000000   133.000000    131.000000        138.0
mean       6.559286     3.390977      4.449962          inf
std        6.800940     3.315310      5.769299          NaN
min        0.000000    -3.000000      0.000000          0.1
25%        5.100000     2.800000      1.600000          0.3
50%        5.750000     3.000000      4.500000          1.3
75%        6.400000     3.300000      5.100000          1.8
max       73.000000    30.000000     63.000000          inf

**suspect found in describe too\  
1)some inf value\
2)-3 and 0 found in min**

In [4]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  140 non-null    float64
 1   Sepal.Width   133 non-null    float64
 2   Petal.Length  131 non-null    float64
 3   Petal.Width   138 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


**suspect found at total entries 150 while column have below 150**

In [16]:
data1.isnull().sum()   # or data1.isna().sum() 

Sepal.Length    10
Sepal.Width     17
Petal.Length    19
Petal.Width     12
Species          0
dtype: int64

**since all null value is already converted to nan which pandas recognize i.e null + entries = total entries**

## 2 Handling missing values
**1 drop--> disadv lose info\
  2.statiscs method-->fill with mean , median , mode**

### 1. drop

In [52]:
data1

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
2             6.2          NaN           5.4          2.3   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
..            ...          ...           ...          ...         ...
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa
148           6.4          3.1           NaN          1.8   virginica
149           5.8          2.6           4.0          NaN  versicolor

[150 rows x 5 columns]

In [53]:
data1.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    140.000000   133.000000    131.000000   137.000000
mean       6.559286     3.390977      4.449962     1.207299
std        6.800940     3.315310      5.769299     0.764722
min        0.000000    -3.000000      0.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.750000     3.000000      4.500000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max       73.000000    30.000000     63.000000     2.500000

**since it contain missing and inf value so replace inf with NaN**

In [54]:
data3 = data1.replace([np.inf, -np.inf], np.nan )


In [55]:
data3 = data3.dropna(how='any')
data3

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
7             5.9          3.0           5.1          1.8   virginica
..            ...          ...           ...          ...         ...
143           6.3          2.9           5.6          1.8   virginica
144           5.7          2.5           5.0          2.0   virginica
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa

[95 rows x 5 columns]

**drop all row having any NaN value and\
need to drop having length negative**

In [62]:
data3 = data3[data3['Sepal.Width']>=0]
data3

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
7             5.9          3.0           5.1          1.8   virginica
..            ...          ...           ...          ...         ...
143           6.3          2.9           5.6          1.8   virginica
144           5.7          2.5           5.0          2.0   virginica
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa

[94 rows x 5 columns]

In [63]:
data3.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0           6.4          3.2           4.5          1.5  versicolor
1           6.3          3.3           6.0          2.5   virginica
3           5.0          3.4           1.6          0.4      setosa
4           5.7          2.6           3.5          1.0  versicolor
7           5.9          3.0           5.1          1.8   virginica

In [64]:
data3.tail()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
143           6.3          2.9           5.6          1.8   virginica
144           5.7          2.5           5.0          2.0   virginica
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa

In [65]:
data3.info()  #satisfied

<class 'pandas.core.frame.DataFrame'>
Index: 94 entries, 0 to 147
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  94 non-null     float64
 1   Sepal.Width   94 non-null     float64
 2   Petal.Length  94 non-null     float64
 3   Petal.Width   94 non-null     float64
 4   Species       94 non-null     object 
dtypes: float64(4), object(1)
memory usage: 4.4+ KB


**since it contain inf value in some row which lead min to infinity**

In [66]:
data3.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count     94.000000    94.000000     94.000000    94.000000
mean       6.346809     3.304255      4.087234     1.200000
std        4.526949     2.828918      2.888957     0.770142
min        4.300000     0.000000      0.000000     0.100000
25%        5.200000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.450000     1.400000
75%        6.500000     3.300000      5.175000     1.875000
max       49.000000    30.000000     23.000000     2.500000

### 2. Statistics

In [77]:
data1

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
2             6.2          NaN           5.4          2.3   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
..            ...          ...           ...          ...         ...
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa
148           6.4          3.1           NaN          1.8   virginica
149           5.8          2.6           4.0          NaN  versicolor

[150 rows x 5 columns]

In [69]:
data1.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    140.000000   133.000000    131.000000   137.000000
mean       6.559286     3.390977      4.449962     1.207299
std        6.800940     3.315310      5.769299     0.764722
min        0.000000    -3.000000      0.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.750000     3.000000      4.500000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max       73.000000    30.000000     63.000000     2.500000

In [71]:
data4 = data1.replace([np.inf, -np.inf, -3], np.nan )
data4

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
2             6.2          NaN           5.4          2.3   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
..            ...          ...           ...          ...         ...
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa
148           6.4          3.1           NaN          1.8   virginica
149           5.8          2.6           4.0          NaN  versicolor

[150 rows x 5 columns]

In [73]:
data4.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    140.000000   132.000000    131.000000   137.000000
mean       6.559286     3.439394      4.449962     1.207299
std        6.800940     3.280401      5.769299     0.764722
min        0.000000     0.000000      0.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.750000     3.000000      4.500000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max       73.000000    30.000000     63.000000     2.500000

In [75]:
s_lm = data4['Sepal.Length'].median()
data4['Sepal.Length'].fillna(value= s_lm , inplace =True)

s_wm = data4['Sepal.Width'].median()
data4['Sepal.Width'].fillna(value = s_wm , inplace =True)

p_lm = data4['Petal.Length'].median()
data4['Petal.Length'].fillna(value = s_wm , inplace =True)

p_wm = data4['Petal.Width'].median()
data4['Petal.Width'].fillna(value = s_wm , inplace =True)

In [76]:
data4

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width     Species
0             6.4          3.2           4.5          1.5  versicolor
1             6.3          3.3           6.0          2.5   virginica
2             6.2          3.0           5.4          2.3   virginica
3             5.0          3.4           1.6          0.4      setosa
4             5.7          2.6           3.5          1.0  versicolor
..            ...          ...           ...          ...         ...
145           6.7          3.1           5.6          2.4   virginica
146           5.6          3.0           4.5          1.5  versicolor
147           5.2          3.5           1.5          0.2      setosa
148           6.4          3.1           3.0          1.8   virginica
149           5.8          2.6           4.0          3.0  versicolor

[150 rows x 5 columns]

In [78]:
data4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  150 non-null    float64
 1   Sepal.Width   150 non-null    float64
 2   Petal.Length  150 non-null    float64
 3   Petal.Width   150 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [79]:
data4.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    150.000000   150.000000    150.000000   150.000000
mean       6.505333     3.386667      4.266300     1.362667
std        6.571880     3.079215      5.410598     0.888747
min        0.000000     0.000000      0.000000     0.100000
25%        5.100000     2.800000      1.700000     0.400000
50%        5.750000     3.000000      4.100000     1.400000
75%        6.400000     3.275000      5.100000     2.000000
max       73.000000    30.000000     63.000000     3.000000